### ISSUE TITLE
[Pass Bug] OutlineIncoreScopes rewrites a pure-Out param to InOut, creating a spurious cross-rank dependency

### ISSUE THREAD

**Issue by zhangqi-chen** [hw-native-sys/pypto#2415]
## Background

In pypto-lib, `models/deepseek_v4_flash_mtp/lm_head.py` on `a2a3` has a **hang**
problem: a 2-card (TP=2) run wedges on the first cross-rank rendezvous and the
scheduler reports no forward progress.

The **root cause is a pass bug, visible compile-only with no device**:
`OutlineIncoreScopes` derives `pl.InOut` for a parameter that is purely written.
`sampled_ids` is declared `pl.Out`, the generated kernel contains a single
`TSTORE` to it and **no load at all**, yet the outlined callee comes out `InOut`
-- and that one wrong direction is what hangs the model.

Reproduce the pass bug with the 35-line case in "Before/Actual IR" below, or on
the model with:

    python models/deepseek_v4_flash_mtp/lm_head.py -p a2a3 --tp 2 --dp 1 --compile-only --dump-passes

Repro case for the hang: `models/deepseek_v4_flash_mtp/lm_head.py`, unmodified
at pypto-lib `main` (`8bf8860f`).

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `8bf8860f` (branch: `main`) |
| pypto | `488ba5d8` (branch: detached) |
| simpler | `1f27a157` (branch: detached) |
| ptoas | `0.57` |
| pto-isa | `f51c92f6` |
| CANN | `9.0.0` |

All four components match their pins (`runtime` gitlink, `pto_isa.pin`,
`toolchain/versions.env`).

Diagnosis: **pypto** -- `OutlineIncoreScopes` derives the wrong argument direction.

### The fix: one tag

The wrong `InOut` propagates into the generated `host_orch.py`, because
`DistributedCodegen::EmitCallToWorker` tags each per-rank chip dispatch argument
from the *callee's* direction:

```python
tensors["t__tmp_v5"] = tensors["sampled_ids__ssa_v0"][r__idx_v0, 0:8, 0:8]
_ta_0.add_tensor(make_tensor_arg(orch._worker, tensors["t__tmp_v5"]), TensorArgType.INOUT)
```

**Changing that single `TensorArgType.INOUT` to `TensorArgType.OUTPUT_EXISTING`
in the generated `host_orch.py` fixes the model.** Editing only that line in a
compiled work dir and replaying it with `--runtime-dir` -- same device binaries,
nothing else touched -- turns the hang into a full pass with golden validation:

```
# unmodified (TensorArgType.INOUT)
PTO2 scheduler timeout sub_class=S1:running-stalled (detail=1) completed=2/9 ...

# only line 33 changed to TensorArgType.OUTPUT_EXISTING
[RUN] runtime done (4.67s)
[RUN] validate done (0.01s)
[RUN] PASS (10.08s)
```

    python models/deepseek_v4_flash_mtp/lm_head.py -p a2a3 --tp 2 --dp 1 -d <c0>,<c1> \
        --runtime-dir build_output/_jit_l3_lm_head_<ts>

`logits` and `sampled_ids` both validate against the torch golden, so the tag
change is not merely hiding the deadlock. The same A/B on the frozen generated
code driven through simpler alone gives the matching picture:

| Variant | Result |
|---|---|
| as generated (`sampled_ids` tagged `INOUT`) | **HANG** |
| identical, `sampled_ids` tagged `OUTPUT_EXISTING` | **PASS** |
| only `sampled_ids` shared across ranks, `INOUT` | **HANG** |
| only `sampled_ids` shared across ranks, `OUTPUT_EXISTING` | **PASS** |
| only `IN`-tagged tensors shared across ranks | PASS |
| per-rank separate storage (no sharing) | PASS |
| sliced with a nonzero offset but per-rank storage | PASS |

Neither the sharing nor the slice offset matters -- only the write-side tag.

### Why the wrong tag deadlocks

The L3 formal stays `pl.Out[[2, 8, 8]]`: one rank-major tensor with rank `r`
writing only row `r`, so the per-rank slices are **disjoint**. With the `INOUT`
tag, both ranks' slices resolve to the same canonical buffer identity under
simpler's address-free Buffer ABI, and a **cross-rank write dependency** is
created between the two chip dispatches. One rank is then held behind the other
while that other rank's kernel spins on the notify the held-back dispatch can
never send:

```
sub_class=S1:running-stalled (detail=1) completed=2/9 running=1 ready=0 waiting=6
orch_done=1 stuck_task_id=4294967298 (ring=1, local=2) stuck_core=50
```

`local=2` is `lm_head_dispatch_wait`, spinning on the peer's `hidden_done` flag
after `lm_head_dispatch_push` completed.

This is a **latent bug that simpler recently exposed, not a simpler
regression**. The same program passes on the pre-migration toolchain (pypto
`b10cae3d` + simpler `3165cc89` + pto-isa `83d01313`), and its `host_orch.py`
emits the **same** `INOUT` tag -- the direction has been wrong all along. It was
inert only because `INOUT` did not yet track dependencies correctly on the
runtime side; now that simpler honours `INOUT` properly, the wrong direction
finally produces the edge it always asked for.

Related: hw-native-sys/pypto#2058 (the opposite direction -- *adding* `InOut` to
recover missing WAR edges).

## Pass Name

`OutlineIncoreScopes`

## Description

When `OutlineIncoreScopes` outlines an incore scope that writes a `pl.Out`
parameter via `pl.tensor.assemble`, it rewrites the outlined function's return
from the assembled value to the incoming parameter, and consequently promotes
that parameter from `Out` to `InOut`. The caller and the top-level formal both
keep `Out`, so the direction becomes inconsistent across the boundary the pass
just introduced.

`InOut` is not justified here: the surrounding `pl.spmd(ROWS)` loop covers every
row, so nothing of the incoming value survives — the parameter is write-only.
The generated device kernel confirms it: `lm_head_greedy_sample.cpp` contains a
single `TSTORE` to `sampled_ids` and **no load at all**.

Downstream, `DistributedCodegen::EmitCallToWorker` tags each per-rank chip
dispatch argument with `ParamDirectionToTensorArgType(callee->param_directions_[i])`
— the callee's (now wrong) `InOut` — which is what creates the spurious
cross-rank dependency described in the Background.

## Git Commit ID

`488ba5d87f382b4bf9b3bc811fab6bdb028149e3`

## Before IR (Input)

`passes_dump/08_after_OutlineHierarchyScopes.py` — parameter is `pl.Out`, the
function returns the assembled value:

```python
@pl.program
class Before:
    @pl.function
    def outline_inout_repro(
        self,
        src__ssa_v0: pl.Tensor[[8, 8], pl.INT32],
        dst__ssa_v0: pl.Out[pl.Tensor[[8, 8], pl.INT32]],
    ) -> pl.Tensor[[8, 8], pl.INT32]:
        for row_inline0__ssa_v0 in pl.spmd(8, name_hint="fill_rows"):
            t__tmp_v0 = pl.tensor.slice(src__ssa_v0, [1, 8], [row_inline0__ssa_v0, 0], [], [])
            dst__ssa_v1: pl.Tensor[[8, 8], pl.INT32] = pl.tensor.assemble(
                dst__ssa_v0, t__tmp_v0, [row_inline0__ssa_v0, 0]
            )
        return dst__ssa_v1
```

## Expected IR (After Transformation)

The outlined callee should keep `pl.Out` — the scope writes every row and reads
none of the incoming value:

```python
@pl.program
class Expected:
    @pl.function
    def fill_rows(
        src__ssa_v0: pl.Tensor[[8, 8], pl.INT32],
        dst__ssa_v0: pl.Out[pl.Tensor[[8, 8], pl.INT32]],
    ) -> pl.Tensor[[8, 8], pl.INT32]:
        t__tmp_v0 = pl.tensor.slice(src__ssa_v0, [1, 8], [row_inline0__ssa_v0, 0], [], [])
        dst__ssa_v1: pl.Tensor[[8, 8], pl.INT32] = pl.tensor.assemble(
            dst__ssa_v0, t__tmp_v0, [row_inline0__ssa_v0, 0]
        )
        return dst__ssa_v1

    @pl.function
    def outline_inout_repro(
        self,
        src__ssa_v0: pl.Tensor[[8, 8], pl.INT32],
        dst__ssa_v0: pl.Out[pl.Tensor[[8, 8], pl.INT32]],
    ) -> pl.Tensor[[8, 8], pl.INT32]:
        for row_inline0__ssa_v0 in pl.spmd(8, name_hint="fill_rows"):
            dst__ssa_v1 = self.fill_rows(src__ssa_v0, dst__ssa_v0)
        return dst__ssa_v1
```

If in-placing the return is deliberate, the direction must at minimum stay
consistent with the caller and the top-level formal, so that a consumer of
`param_directions_` cannot infer a read that never happens.

## Actual IR or Error

`passes_dump/09_after_OutlineIncoreScopes.py` — callee is `pl.InOut`, the return
is rewritten to the incoming parameter, and the caller still says `pl.Out`:

```python
def fill_rows(src__ssa_v0: pl.Tensor[[8, 8], pl.INT32],
              dst__ssa_v0: pl.InOut[pl.Tensor[[8, 8], pl.INT32]]) -> pl.Tensor[[8, 8], pl.INT32]:
    t__tmp_v0 = pl.tensor.slice(src__ssa_v0, [1, 8], [row_inline0__ssa_v0, 0], [], [])
    dst__ssa_v1: pl.Tensor[[8, 8], pl.INT32] = pl.tensor.assemble(dst__ssa_v0, t__tmp_v0, [row_inline0__ssa_v0, 0])
    return dst__ssa_v0

def outline_inout_repro(self, src__ssa_v0: pl.Tensor[[8, 8], pl.INT32],
                        dst__ssa_v0: pl.Out[pl.Tensor[[8, 8], pl.INT32]]) -> pl.Tensor[[8, 8], pl.INT32]:
    for row_inline0__ssa_v0 in pl.spmd(8, name_hint="fill_rows"):
        dst__ssa_v1: pl.Tensor[[8, 8], pl.INT32] = self.fill_rows(src__ssa_v0, dst__ssa_v0)
    return dst__ssa_v1
```

The real model shows the identical rewrite on `sampled_ids`
(`08` line 16 `pl.Out` → `09` line 97 `pl.InOut`, return `v1` → `v0`).

<details>
<summary>Minimal reproduction case — <code>repro_outline_inout.py</code> (not committed; save under pypto-lib and run)</summary>

```python
import pypto.language as pl

ROWS = 8
COLS = 8


@pl.jit.inline
def fill_rows(
    src: pl.Tensor[[ROWS, COLS], pl.INT32],
    dst: pl.Tensor[[ROWS, COLS], pl.INT32],
):
    for row in pl.spmd(ROWS, name_hint="fill_rows"):
        dst[row : row + 1, 0:COLS] = src[row : row + 1, 0:COLS]


@pl.jit
def outline_inout_repro(
    src: pl.Tensor[[ROWS, COLS], pl.INT32],
    dst: pl.Out[pl.Tensor[[ROWS, COLS], pl.INT32]],
):
    fill_rows(src, dst)
    return dst


def build_specs():
    import torch
    from golden import TensorSpec

    return [
        TensorSpec("src", shape=[ROWS, COLS], dtype=torch.int32, init_value=torch.ones),
        TensorSpec("dst", shape=[ROWS, COLS], dtype=torch.int32, is_output=True),
    ]


def golden_outline_inout_repro(src, dst):
    return src.clone()


if __name__ == "__main__":
    import argparse
    import sys

    sys.path.insert(0, ".")
    from golden import run_jit

    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", type=str, default="a2a3")
    args = parser.parse_args()

    run_jit(
        fn=outline_inout_repro,
        specs=build_specs(),
        golden_fn=golden_outline_inout_repro,
        runtime_cfg=dict(platform=args.platform, device_id=0),
        compile_cfg=dict(dump_passes=True),
        compile_only=True,
    )
```

Run it from the pypto-lib root (compile-only, no device needed), then compare
`dst` in `build_output/_jit_outline_inout_repro_*/passes_dump/08_after_OutlineHierarchyScopes.py`
against `09_after_OutlineIncoreScopes.py`.

</details>

## Attached build output

`pypto-2415-build-output.tar.gz` (attached below) carries the three work dirs
this report cites. Absolute paths in it are replaced with placeholders
(`<pypto-lib>`, `<pypto>`, `<pto-isa>`, `<workdir>`); nothing else is modified.

| Directory | What it is | Where to look |
|---|---|---|
| `_jit_l3_lm_head_20260818_022626/` | `lm_head` compiled on the environment above, `--compile-only --dump-passes` | `passes_dump/08_after_OutlineHierarchyScopes.py` line 16 (`pl.Out`) vs `09_after_OutlineIncoreScopes.py` line 97 (`pl.InOut`, return rewritten `v1` -> `v0`); `orchestration/host_orch.py` line 33 (the `TensorArgType.INOUT` to flip); `next_levels/lm_head_with_sampling_test/kernels/aiv/lm_head_greedy_sample.cpp` (single `TSTORE`, no load); `kernel_config.py` line 19 (`_D.INOUT` in the orchestration signature) |
| `_jit_outline_inout_repro_20260818_024007/` | the 35-line minimal case above, same flags | same two `passes_dump` files -- `dst` goes `pl.Out` -> `pl.InOut` with 4 lines of IR |
| `_jit_l3_lm_head_20260818_022435/` | the same `lm_head` compiled on the pre-migration toolchain (pypto `b10cae3d` + simpler `3165cc89` + pto-isa `83d01313`) | `orchestration/host_orch.py` -- emits the **same** `TensorArgType.INOUT` on line 33, showing the direction was already wrong before the runtime started honouring it |

## NPU Kind (if relevant)

Ascend 910B

## Host Platform (if relevant)

Linux (aarch64)

**Comment by zhangqi-chen** [comment]
[pypto-2415-build-output.tar.gz](https://github.com/user-attachments/files/31175384/pypto-2415-build-output.tar.gz)
